src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -20,11 +20,17 @@ namespace System.IO
 {
     public partial class FileSystemWatcher
     {
+        private GCHandle _cachedRunningInstance;
```

Comment:
Should this GCHandle rather be in the `RunningInstance` ? `RunningInstance` has the callback that the GCHandle protects, it would be best if this GCHandle is close to where the protected callbacks are.

Reply:
You can actually simplify the RunningInstance to just use function pointer + GCHandle once you do that. The marshaled delegate that it uses right now is less efficient than function pointer + GCHandle.